config/experimental/auth/auth.go

```diff
@@ -0,0 +1,218 @@
+// handle retries appropriately.
+//
+// If the TokenSource is already a cached token source (obtained by calling this
+// function), it is returned as is.
+func NewCachedTokenSource(ts oauth2.TokenSource, opts ...Option) oauth2.TokenSource {
+	if cts, ok := ts.(*cachedTokenSource); ok {
+		return cts
+	}
+
+	cts := &cachedTokenSource{
+		tokenSource:   ts,
+		staleDuration: defaultStaleDuration,
+		disableAsync:  defaultDisableAsyncRefresh,
+		cachedToken:   nil,
+		timeNow:       time.Now,
+	}
+
+	for _, opt := range opts {
+		opt(cts)
+	}
+
+	return cts
+}
+
+type cachedTokenSource struct {
+	// The token source to obtain tokens from.
+	tokenSource oauth2.TokenSource
+
+	// If true, only refresh the token with a blocking call when it is expired.
+	disableAsync bool
+
+	// Duration during which a token is considered stale, see tokenState.
+	staleDuration time.Duration
+
+	mu          sync.Mutex
+	cachedToken *oauth2.Token
+
+	// Indicates that an async refresh is in progress. This is used to prevent
+	// multiple async refreshes from being triggered at the same time.
+	isRefreshing bool
+
+	// Error returned by the last refresh. Async refreshes are disabled if this
+	// value is not nil so that the cache does not continue sending request to
+	// a potentially failing server. The next blocking call will re-enable async
+	// refreshes by setting this value to nil if it succeeds, or return the
+	// error if it fails.
+	refreshErr error
+
+	timeNow func() time.Time // for testing
+}
+
+// Token returns a token from the cache or fetches a new one if the current
+// token is expired.
+func (cts *cachedTokenSource) Token() (*oauth2.Token, error) {
+	if cts.disableAsync {
+		return cts.blockingToken()
+	}
+	return cts.asyncToken()
+}
+
+// tokenState represents the state of the token. Each token can be in one of
+// the following three states:
+//   - fresh: The token is valid.
+//   - stale: The token is valid but will expire soon.
+//   - expired: The token has expired and cannot be used.
+//
+// Token state through time:
+//
+//	issue time     expiry time
+//	    v               v
+//	    | fresh | stale | expired -> time
+//	    |     valid     |
+type tokenState int
+
+const (
+	fresh   tokenState = iota // The token is valid.
+	stale                     // The token is valid but will expire soon.
+	expired                   // The token has expired and cannot be used.
+)
+
+// tokenState returns the state of the token. The function is not thread-safe
+// and should be called with the lock held.
+func (c *cachedTokenSource) tokenState() tokenState {
+	if c.cachedToken == nil {
+		return expired
+	}
+	switch lifeSpan := c.cachedToken.Expiry.Sub(c.timeNow()); {
+	case lifeSpan <= 0:
+		return expired
+	case lifeSpan <= c.staleDuration:
+		return stale
+	default:
+		return fresh
+	}
+}
+
+func (cts *cachedTokenSource) asyncToken() (*oauth2.Token, error) {
+	cts.mu.Lock()
+	ts := cts.tokenState()
+	t := cts.cachedToken
+	cts.mu.Unlock()
```

Comment:
nit: maybe a helper function that returns both tokenState() and cachedToken and takes care of locking internally would be useful there, because it would allow you to switch to 'defer cts.mu.Unlock()' so that you know that mutex is unlocked on panics as well.